### PR TITLE
fix(visualsign): reject JSON short-form control escapes in validate_charset

### PR DESCRIPTION
**Pull request by pepe-anchor**
## Summary

`SignablePayload::validate_charset()` was only blocking `\u` substrings in the serialized JSON, then leaning on `is_ascii() && (is_ascii_graphic() || is_ascii_whitespace())`. `serde_json::CompactFormatter` emits short-form escapes (`\b`, `\t`, `\n`, `\f`, `\r`, `\"`, `\\`) for control bytes and structural chars; all of those (except `\n`, see carve-out below) slip past the ASCII checks while still smuggling real control bytes into the decoded display string. The Tron converter forwarding `parameter.type_url` straight into a text field is the concrete sink; an attacker can break single-line wallet UI or splice display lines via `\r`, `\t`, etc.

## What changed

- `src/visualsign/src/lib.rs::validate_charset`: replaced the single `"\\u"` substring check with a list covering the escapes the serializer can emit that we want to reject: `\u`, `\t`, `\r`, `\b`, `\f`, `\"`, `\\`, plus `\/` defensively (the compact formatter does not currently emit `\/`, but the rule documents the intent in case a future serializer or replacement ever does). `\/` is listed ahead of `\\` so the defensive guard actually fires first on any future input containing bare `\/`.
- **Newline carve-out**: `\n` is intentionally **not** in the deny-list. It is the wallet's documented line separator for multi-line field text (e.g. "Program ID: X\nData: Y") and 30+ trusted call sites across the chain converters emit it deliberately. Untrusted strings flowing into field text must be sanitized at the insertion site instead. `test_validate_charset_accepts_newline_in_text` pins this contract.
- Added regression tests in the same file:
  - `test_validate_charset_rejects_short_form_control_escapes` (4 cases: `\t`, `\r`, `\b`, `\f`; each pins the exact short-form escape the serializer emits)
  - `test_validate_charset_rejects_quote_and_backslash_escapes`
  - `test_validate_charset_rejects_backslash_slash_combination`
  - `test_validate_charset_still_rejects_unicode_escape`
  - `test_validate_charset_accepts_plain_ascii` (positive case)
  - `test_validate_charset_accepts_newline_in_text` (positive case, pins carve-out)

No callers changed. The fix is purely an extra deny-list inside the validator, so any payload that already passed continues to pass; payloads that snuck control bytes through (other than `\n`) now get rejected with the same `ValidationError("Restricted Characters Detected")` already used for `\u`.

## Rollback

Revert this commit. No data migrations, no protobuf changes, no on-disk state. The validator returns to its previous (looser) behavior; nothing downstream depends on the new rejections beyond the new tests in this PR.

## Backwards compatibility

Strictly tightens validation on data that should never have been allowed past it. No public API changed. Any caller currently producing payloads that contain `"`, `\`, or control bytes other than `\n` in text fields was already producing JSON that broke the validator's intent; those callers need to sanitize their input. None of the existing chain converters in this repo exercise that path with the current fixture corpus (visualsign unit + doctest suite is green; cross-chain fixture verification is left for the PR-level CI per scope guidance on the ticket).</p>

## Test plan

Verification scoped to `visualsign`, ran from `src/`:

- [x] `cargo fmt`, no diff
- [x] `cargo clippy -p visualsign --all-targets -- -D warnings`, clean
- [x] `cargo test -p visualsign`, all passing (incl. new regression tests and the pre-existing `test_signable_payload_to_json` / charset suite)

Cross-chain CI on this PR will exercise the remaining converters end-to-end.

### CODE REVIEW

File: src/visualsign/src/lib.rs [hunk 1]
```diff
@@ -904,15 +904,40 @@ impl SignablePayload {
...
+        //
+        // Carve-out: `\n` is the wallet's documented line separator for
+        // multi-line field text (e.g. "Program ID: X\nData: Y" used by every
+        // chain parser's instruction display). 30+ trusted call sites across
+        // the chain converters emit it deliberately. Rejecting it here would
+        // break legitimate output. Attacker-controlled strings destined for
+        // field text must instead be sanitized at the insertion site (PRS-231
+        // follow-up tracks the remaining sinks, e.g. Tron `parameter.type_url`).
+        const FORBIDDEN_JSON_ESCAPES: &[&str] =
+            &["\\u", "\\t", "\\r", "\\b", "\\f", "\\\"", "\\\\", "\\/"];
```
[the remark this answers is not in the document]
Comment (pepe-anchor) [reply src/visualsign/src/lib.rs]: Good catch. Reordered FORBIDDEN_JSON_ESCAPES so `\/` is checked before `\\` in 1dfed6c. Now the `\/` rule fires first on any input containing the substring `\/` (including the `\\/` produced by a literal backslash in front of a `/`), so the defensive entry actually attributes the rejection rather than being shadowed.

I kept it as a substring check rather than adding a separate test that hits `\/` without `\\`, since the only way to inject bare `\/` into serialized JSON is to bypass `serde_json::CompactFormatter` entirely (which doesn't emit it). The reordering + the existing backslash-slash test exercise the rule end-to-end with the real serializer.

If you disagree, react with 👎 and we'll revisit.

File: src/visualsign/src/lib.rs [hunk 2]
```diff
@@ -2654,4 +2679,148 @@ mod tests {
...
+        // The simplest way to inject the literal substring `\/` into the
+        // serialized JSON is to put a backslash directly in front of a
+        // forward slash in field text. The `\` byte is itself escaped as
+        // `\\`, producing `\\/` in the serialized JSON; the validator's
+        // backslash rule catches that as well. This documents that any
+        // path producing `\/` (or `\\`) is rejected.
+        let payload = payload_with_text("path\\/to/resource");
+        let err = payload
+            .validate_charset()
+            .expect_err("validator must reject escaped slash / backslash combinations");
```
[the remark this answers is not in the document]
Comment (pepe-anchor) [reply src/visualsign/src/lib.rs]: Right, the original test name was misleading. Renamed it to `test_validate_charset_rejects_backslash_slash_combination` in 1dfed6c and documented inline that the real serializer can't emit bare `\/`, so the test exercises the backslash-slash combination (`\\/` in serialized JSON) which now hits the `\/` rule first thanks to the reordering. The doc-comment also points at the companion test that covers the pure `\\` path so the boundary is explicit.

If you disagree, react with 👎 and we'll revisit.

File: src/visualsign/src/lib.rs [hunk 3]
```diff
@@ -904,15 +904,40 @@ impl SignablePayload {
...
+        // wallet UI. `\/` is included defensively even though the compact
+        // formatter does not emit it.
+        //
+        // Carve-out: `\n` is the wallet's documented line separator for
+        // multi-line field text (e.g. "Program ID: X\nData: Y" used by every
+        // chain parser's instruction display). 30+ trusted call sites across
+        // the chain converters emit it deliberately. Rejecting it here would
+        // break legitimate output. Attacker-controlled strings destined for
+        // field text must instead be sanitized at the insertion site (PRS-231
+        // follow-up tracks the remaining sinks, e.g. Tron `parameter.type_url`).
```
[the remark this answers is not in the document]
Comment (pepe-anchor) [reply src/visualsign/src/lib.rs]: Confirmed, the PR body was inaccurate. The implementation intentionally carves `\n` out of the deny-list (it's the wallet's documented multi-line separator, used by every chain converter's instruction display) and 30+ trusted call sites emit it. Updated the PR body to reflect this: deny-list is `\t/\r/\b/\f/\"/\\/\/u` (no `\n`), with the carve-out called out explicitly and the new `test_validate_charset_accepts_newline_in_text` test pinning that contract.

If you disagree, react with 👎 and we'll revisit.

File: src/visualsign/src/lib.rs [hunk 4]
```diff
@@ -2654,4 +2679,148 @@ mod tests {
...
+
+        for (label, text) in cases {
+            let payload = payload_with_text(text);
+            let json = payload.to_json().expect("serialization should succeed");
+            // Sanity-check: serde_json's CompactFormatter does emit the
+            // short-form escape, which is exactly what the old validator
+            // missed.
+            assert!(
+                json.contains('\\'),
+                "{label}: expected serialized JSON to contain a backslash escape, got: {json}",
```
[the remark this answers is not in the document]
Comment (pepe-anchor) [reply src/visualsign/src/lib.rs]: Good point, the generic backslash check would have masked a serializer regression. Tightened the case tuple in 1dfed6c to `(label, text, expected_escape)` and pinned each to its specific short-form escape (`\t`, `\r`, `\b`, `\f`). If serde ever switches to `\uXXXX` long-form, the sanity-check fails per case and we know to revisit (the validator itself would still reject via the `\u` rule, so test failure is the right signal).

If you disagree, react with 👎 and we'll revisit.

File: src/visualsign/src/lib.rs [hunk 5]
```diff
@@ -904,15 +904,46 @@ impl SignablePayload {
...
+        // the chain converters emit it deliberately. Rejecting it here would
+        // break legitimate output. Attacker-controlled strings destined for
+        // field text must instead be sanitized at the insertion site (PRS-231
+        // follow-up tracks the remaining sinks, e.g. Tron `parameter.type_url`).
+        // Order matters: `\/` is listed before `\\` so the defensive `\/`
+        // entry can actually attribute a rejection in the (currently
+        // unreachable with `serde_json::CompactFormatter`) case where a
+        // serializer emits `\/`. With `\\` listed first, any input containing
+        // a literal backslash would hit the `\\` rule before the loop ever
+        // checked `\/`, masking whether the `\/` guard is doing anything.
```
[the remark this answers is not in the document]
Comment (pepe-anchor) [reply src/visualsign/src/lib.rs]: Fair point, and you are right. The loop returns the same generic `ValidationError("Restricted Characters Detected")` for every match, so there's no observable difference between which entry fires first. The "attribution" framing in my previous reply overstated what the ordering actually buys us.

Went with option (a) in 5bd107a: trimmed the long comment on `FORBIDDEN_JSON_ESCAPES` to flag that list order has no runtime effect (defensive-first is cosmetic only), and dropped the parallel claim in `test_validate_charset_rejects_backslash_slash_combination`'s doc-comment. Kept the generic error since not leaking which rule matched is a deliberate security-conscious choice across both this loop and the non-printable path.

If you disagree, react with 👎 and we'll revisit.

File: src/visualsign/src/lib.rs [hunk 6]
```diff
@@ -2654,4 +2683,163 @@ mod tests {
...
+            ("carriage return", "hello\rworld", "\\r"),
+            ("backspace", "hello\u{0008}world", "\\b"),
+            ("form feed", "hello\u{000C}world", "\\f"),
+        ];
+
+        for (label, text, expected_escape) in cases {
+            let payload = payload_with_text(text);
+            let json = payload.to_json().expect("serialization should succeed");
+            // Sanity-check: serde_json's CompactFormatter emits the specific
+            // short-form escape (not e.g. `	`). If it ever switched to
```
[the remark this answers is not in the document]
Comment (pepe-anchor) [reply src/visualsign/src/lib.rs]: Good catch, replaced the literal U+0009 with an explicit `U+0009` reference in the comment so the intent is readable in source and won't get mangled by editors/linters. Fixed in 6db2068.

If you disagree, react with 👎 and we'll revisit.

**Comment by prasanna-anchorage** [comment]
Heads up: this repo is **public**, so internal Linear refs (`PRS-xxx`) shouldn't leak into PR-visible surface — external readers can't resolve them.

This PR's leaks:
- Body links `linear.app/anchorlabs/issue/PRS-231` — swap for a one-line bug-class summary or a public GH issue reference.
- 5 code references / 5 inline comments mention `PRS-231` — rename to describe the bug class.
- Branch `pepefigueira/prs-231-...` also exposes the ID (optional).

Suggested convention: describe the bug in the name (e.g., `test_short_form_control_escapes_rejected`). For persistent traceability, file a public GitHub issue.

**Comment by pepe-anchor** [comment]
Scrubbed PRS-231 refs from body and code/comments. Branch left as-is since it'll disappear on merge.

**Comment by prasanna-anchorage** [comment]
Deep-review verdict: **ship as-is**.

Verified the validator/serde contract empirically against `serde_json::CompactFormatter`:

| Input byte | serde emits | Caught by |
|---|---|---|
| 0x08, 0x09, 0x0C, 0x0D | `\b`, `\t`, `\f`, `\r` | new substring rules ✓ |
| 0x0A LF | `\n` | carve-out (allowed) |
| 0x22 `"`, 0x5C `\` (in string) | `\"`, `\\` | new substring rules ✓ |
| 0x7F DEL | literal byte | pre-existing per-char `is_ascii_graphic`/`is_ascii_whitespace` ✓ |
| U+202E, ZWSP, non-ASCII | literal UTF-8 bytes | pre-existing `!is_ascii()` ✓ |
| 0x2F `/` | literal `/` | `\/` rule is dead defensive code (harmless) |

No emit/validate mismatch — the most important property for a string-pattern validator. Tests inject real control bytes and route them through `to_json()`, which pins the validator-emitter contract so a future serde change can't silently move the goalposts.

**Residual gap (follow-up, not blocking):** the `\n` carve-out is a UI-line-spoof primitive. An attacker who controls a string field can embed `\n\n` to fake a second wallet UI line (`"Approved spender\n\nRecipient: 0xMallory"`). Closing this needs per-sink sanitization, not a stricter central gate — per the in-source comment at lines 922-924. Tracking issue suggested: a `sanitize_user_text(&str) -> String` helper in `visualsign` that strips `\n` for use at every place where caller-controlled metadata flows into a field body.

#331 is now merged so the gate is closed on every chain; this PR makes the validator strict enough to catch the short-form escapes that previously slipped past `is_ascii_whitespace`. Together they're complete coverage modulo the per-sink follow-up above.

**Review by prasanna-anchorage — APPROVED** [review]
Approving per deep-review verdict above — ship as-is.
